File: apps/swift/Sources/PackRat/Features/Chat/ChatPackTools.swift
import Foundation

/// The local-data side of the assistant's pack tools.
///
/// Tools that touch the user's own packs are declared server-side with no
/// `execute` (see `packages/api/src/utils/ai/tools.ts`), so the model's call is
/// streamed down and answered here, against the device's local store.
///
/// That split is deliberate rather than incidental. The local store is what the
/// user is looking at and is also the write path — mutations land there and sync
/// outward through the outbox. Answering from the server instead would mean an
/// item the assistant "added" stayed invisible until the next refresh, and that
/// nothing worked offline.
@MainActor
protocol ChatPackToolHandling {
    /// Packs matching `nameQuery`, or all packs when it is nil/empty.
    func listPacks(nameQuery: String?) -> [ChatPackSummary]

    /// A pack's full contents, or nil when no pack has that id.
    func packDetails(id: String) -> [String: String]?

    /// Adds an item to a pack, returning a short confirmation.
    /// Throws if the pack is unknown or the write fails.
    func addItem(_ request: ChatAddItemRequest) async throws -> ChatAddItemResult
}

/// The subset of a pack the model needs in order to pick one by name.
struct ChatPackSummary: Encodable {
    let id: String
    let name: String
    let category: String?
    let description: String?
}

/// A model-requested item insertion, decoded from the tool call's arguments.
struct ChatAddItemRequest {
    let packId: String
    let name: String
    let weight: Double?
    let weightUnit: WeightUnit?
    let quantity: Int
    let category: String?
    let consumable: Bool
    let worn: Bool
    let notes: String?
    let catalogItemId: String?

    /// Decodes the tool arguments, defaulting the optional fields the way the
    /// tool's schema documents. Returns nil when a required field is missing, so
    /// a malformed call becomes a tool error the model can react to rather than
    /// a silently wrong insert.
    init?(toolArguments args: [String: Any]) {
        guard let packId = args["packId"] as? String, !packId.isEmpty,
              let name = (args["name"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines),
              !name.isEmpty
        else { return nil }

        self.packId = packId
        self.name = name
        // JSON numbers decode as NSNumber, so accept either representation.
        self.weight = (args["weight"] as? Double) ?? (args["weight"] as? Int).map(Double.init)
        self.weightUnit = (args["weightUnit"] as? String).flatMap { WeightUnit(apiValue: $0) }
        self.quantity = max(1, (args["quantity"] as? Int) ?? 1)
        self.category = Self.trimmedString(args["category"])
        self.consumable = (args["consumable"] as? Bool) ?? false
        self.worn = (args["worn"] as? Bool) ?? false
        self.notes = Self.trimmedString(args["notes"])
        self.catalogItemId = Self.trimmedString(args["catalogItemId"])
    }

    /// Models like to send `""` for "not set"; treat that as absent so an empty
    /// string never lands in the store as a category or note.
    private static func trimmedString(_ value: Any?) -> String? {
        guard let text = (value as? String)?.trimmingCharacters(in: .whitespacesAndNewlines),
              !text.isEmpty
        else { return nil }
        return text
    }
}

/// What the model is told after a successful insert.
struct ChatAddItemResult: Encodable {
    let itemId: String
    let itemName: String
    let packId: String
    let packName: String
    let quantity: Int
    /// True when the item was queued locally because the device is offline. The
    /// model should still report success — the write is durable either way — but
    /// this lets it mention the pending sync.
    let pendingSync: Bool
}

/// Errors surfaced to the model as a tool failure.
enum ChatPackToolError: LocalizedError {
    case packNotFound(String)
    case invalidArguments

    var errorDescription: String? {
        switch self {
        case .packNotFound(let id):
            return "No pack with id \(id) exists on this device."
        case .invalidArguments:
            return "The tool call was missing a required field (packId and name are required)."
        }
    }
}

File: apps/expo/features/packs/utils/chatPackTools.ts
/**
 * Client-side implementations of the assistant's pack tools.
 *
 * `listUserPacks` and `addItemToPack` are declared on the server with no
 * `execute` (packages/api/src/utils/ai/tools.ts), so the model's call is
 * streamed to the client and answered from the local store. That keeps reads
 * consistent with what the user is looking at and keeps writes on the offline
 * path — but it means every client MUST answer both tools. A client that
 * ignores one leaves the tool call unresolved and the turn stalls forever.
 *
 * These helpers hold the shared logic so the remote chat screen and the
 * on-device toolset cannot drift apart.
 */

import type { PackInStore, PackItem, PackItemInput, WeightUnit } from '../types';
import { matchPacksByName, type PackMatchSummary } from './matchPacksByName';

export type ListUserPacksInput = { nameQuery?: string | null };

export type AddItemToPackInput = {
  packId: string;
  name: string;
  weight?: number;
  weightUnit?: WeightUnit;
  quantity?: number;
  category?: string;
  consumable?: boolean;
  worn?: boolean;
  notes?: string;
  catalogItemId?: string;
};

export type ToolResult<T> = { success: true; data: T } | { success: false; error: string };

/**
 * Resolve a pack name the user mentioned into pack ids the model can use.
 *
 * Returns the match list rather than a single pack: when a query matches
 * several packs the model needs to see them all so it can ask which one,
 * instead of silently writing to the wrong pack.
 */
export function listUserPacksFromStore({
  packs,
  nameQuery,
}: {
  packs: Record<string, PackInStore>;
  nameQuery?: string | null;
}): ToolResult<PackMatchSummary[]> {
  const matches = matchPacksByName({ packs: Object.values(packs), nameQuery });
  return { success: true, data: matches };
}

/** Catalog ids in the API are numeric; a non-numeric value from the model means
 *  "not a catalog item" rather than an error. */
function parseCatalogItemId(value: string | undefined): number | undefined {
  if (value === undefined) return undefined;
  const parsed = Number(value);
  return Number.isInteger(parsed) && parsed > 0 ? parsed : undefined;
}

/**
 * Build the pack-item payload for an `addItemToPack` call.
 *
 * Split out from the write itself so the defaulting rules are unit testable
 * without a Legend-State store: an item added without a weight silently
 * degrades the pack totals the app exists to compute, so the defaults matter.
 */
export function buildPackItemInput(input: AddItemToPackInput): PackItemInput {
  return {
    name: input.name.trim(),
    weight: input.weight ?? 0,
    weightUnit: input.weightUnit ?? 'g',
    quantity: Math.max(1, Math.floor(input.quantity ?? 1)),
    category: input.category?.trim() || 'general',
    consumable: input.consumable ?? false,
    worn: input.worn ?? false,
    notes: input.notes,
    catalogItemId: parseCatalogItemId(input.catalogItemId),
  };
}

/**
 * Validate an `addItemToPack` call against the local store and produce the item
 * payload to write.
 *
 * Returns the resolved pack so the caller can confirm which pack it wrote to —
 * the assistant has to name the pack back to the user, and it must not claim
 * success against a pack that is deleted or was never there.
 */
export function prepareAddItemToPack({
  packs,
  input,
}: {
  packs: Record<string, PackInStore>;
  input: AddItemToPackInput;
}): ToolResult<{ pack: PackInStore; itemData: PackItemInput }> {
  const pack = packs[input.packId];
  if (!pack || pack.deleted) {
    return {
      success: false,
      error: `No pack with id ${input.packId} exists on this device. Call listUserPacks to resolve the pack by name first.`,
    };
  }

  if (!input.name.trim()) {
    return { success: false, error: 'An item name is required.' };
  }

  return { success: true, data: { pack, itemData: buildPackItemInput(input) } };
}

/** Shape the tool output the model sees after a successful add. */
export function describeAddedItem({
  pack,
  item,
}: {
  pack: PackInStore;
  item: Pick<PackItem, 'id' | 'name' | 'quantity' | 'weight' | 'weightUnit' | 'category'>;
}) {
  return {
    packId: pack.id,
    packName: pack.name,
    item: {
      id: item.id,
      name: item.name,
      quantity: item.quantity,
      weight: item.weight,
      weightUnit: item.weightUnit,
      category: item.category,
    },
  };
}
